src/linear_algebra/free_module/strong_rank_condition.lean

```diff
@@ -0,0 +1,51 @@
+/-
+Copyright (c) 2021 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import linear_algebra.charpoly.basic
+import linear_algebra.invariant_basis_number
+
+/-!
+
+# Strong rank condition for commutative rings
+
+We prove that any nontrivial commutative ring satisfies `strong_rank_condition`.
+
+## Main result
+
+* `comm_ring_strong_rank_condition R` : `R` has the `strong_rank_condition`.
+
+## References
+
+We follow the proof given in https://mathoverflow.net/a/47846/7845.
+
+-/
+
+variables (R : Type*) [comm_ring R] [nontrivial R]
+
+open polynomial function fin linear_map
+
+instance comm_ring_strong_rank_condition : strong_rank_condition R :=
+begin
+  suffices : ∀ n, ∀ f : (fin (n + 1) → R) →ₗ[R] fin n → R, ¬injective f,
+  { rwa strong_rank_condition_iff_succ R },
+  intros n f, by_contradiction hf,
+
+  letI := module.finite.of_basis (pi.basis_fun R (fin (n + 1))),
+
+  let g : (fin (n + 1) → R) →ₗ[R] fin (n + 1) → R :=
+    (extend_by_zero.linear_map R cast_succ).comp f,
+  have hg : injective g := (extend_injective (rel_embedding.injective cast_succ) 0).comp hf,
+
+  have hnex : ¬∃ i : fin n, cast_succ i = last n := λ ⟨i, hi⟩, ne_of_lt (cast_succ_lt_last i) hi,
+
+-- Since `g` is injective, its minimal polynomial `P` has nonzero constant term `a₀`, but evaluating
+-- `P(g)` at the vector `(0,...,0,1)` gives `a₀ = 0`, a contradiction.
+  let heval := minpoly.aeval R g,
+  obtain ⟨P₁, hP₁⟩ := X_dvd_iff.2 (erase_same (minpoly R g) 0),
+  rw [← monomial_add_erase (minpoly R g) 0, hP₁, linear_map.ext_iff] at heval,
+  replace heval := congr_fun (heval (pi.single (last n) 1)) (last n),
+  simpa [hnex, charpoly_coeff_zero_of_injective hg] using heval
```

Comment:
Here's an attempt at leanifying this comment:
```lean
  let a₀ := (minpoly R g).coeff 0,
  have : a₀ ≠ 0 := charpoly_coeff_zero_of_injective hg,
  have : a₀ = 0,
  { -- evaluate the minimal polynomial at the vector `(0,...,0,1)`
    have heval := linear_map.congr_fun (minpoly.aeval R g) (pi.single (fin.last n) 1),
    obtain ⟨P, hP⟩ := X_dvd_iff.2 (erase_same (minpoly R g) 0),
    rw [← monomial_add_erase (minpoly R g) 0, hP] at heval,
    replace heval := congr_fun heval (fin.last n),
    simpa [hnex] using heval },
  contradiction,
```

Reply:
Yes, at the end the proof is really simple. I wanted to explain it with some comments because it is probably quite surprising (even if the result is well known, the standard proofs are far more complicated).

Reply:
Sorry, but I have very strong feelings about this: more comments are *always* good, so one should probably never ask to remove comments. (And one should probably ask more often to add comments, by the way :-)

Reply:
Note that I didn't ask for the comment to be removed, but rather for the lean code to be written in a way that makes it redundant. Surely you would advocate for removal of a comment in code like
```lean
-- let x be 3
let x := 3
```

Reply:
Of course in your example the comment doesn't make any sense. IMHO, In Riccardo's situation it does make sense: reading a sentence that insists on the articulation of the arguments is more explanatory than reading a few lines of code giving a juxtaposition of statements.

Reply:
Would you prefer the original, or the new version with the original comment?

Reply:
The new version is much clearer to me, so I'd go for the new version with the original comment. Of course, all this is pretty subjective, so in the end I'll let Riccardo decide what he likes best!

Reply:
I agree this new version is better!